src/lib/input/input.ts

```diff
@@ -90,14 +95,46 @@ export class MdHint {
   @Input() align: 'start' | 'end' = 'start';
 }
 
+/**
+ * Enables auto expanding behaviour for the textarea
+ */
+@Directive({
+  selector: 'textarea[mdAutosize]'
+})
+export class MdAutosize implements OnInit {
+  @HostListener('input', ['$event.target'])
+  public onChange() {
+    if (this._isActive) {
+      this._adjust();
+    }
+  }
+
+  @Input('mdAutosize')
+  private _isActive: boolean;
+
+  constructor(private _elRef: ElementRef) {}
+
+  ngOnInit() {
+    if (this._isActive) {
+      this._adjust();
+    }
+  }
+
+  private _adjust() {
+    this._elRef.nativeElement.style.overflow = 'hidden';
+    this._elRef.nativeElement.style.height = 'auto';
+    this._elRef.nativeElement.style.height = `${this._elRef.nativeElement.scrollHeight}px`;
```

Comment:
Correct, most of the complexity is because of the `rows`, however I think that they're a pretty important part of the component, because they're what sets an empty `textarea` apart from an `input`. In Material 1 we also do the following, although I'm not the one who wrote it so I'm not sure what the reasoning is:

```js
function getHeight() {
  var offsetHeight = node.offsetHeight;
  var line = node.scrollHeight - offsetHeight;
  return offsetHeight + Math.max(line, 0);
}
```